lib/core/include/irods/configuration_hook_manager.hpp

```diff
@@ -0,0 +1,117 @@
+#ifndef IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+#define IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+
+/// \file
+
+#include <nlohmann/json.hpp>
+
+#include <vector>
+#include <string>
+#include <unordered_map>
+
+namespace irods::experimental::configuration_reload
+{
+    /// \brief The string used for the catch-all hooks that are called on every configuration change.
+    /// \see hook
+    const std::string CATCHALL_KEYWORD("*CATCHALL*");
+
+    /// \brief Hooks are functions called with the pathname of the json value that has changed, as well as the new
+    /// operation performed, removal versus replacement versus insertion etc.
+    ///
+    /// Hooks are invoked in two stages for each entry in a list of json patches.
+    ///
+    /// The first is the CATCHALL_KEYWORD stage, where hooks that are run for every
+    /// configuration change occurs. The second is based on the specific json
+    /// path of the changed configuration value.
+    ///
+    /// For example, if you change the delay server's sleep duration, the
+    /// CATCHALL_KEYWORD hooks will be run first in order of addition, followed by
+    /// the hooks associated with
+    /// "/advanced_settings/delay_server_sleep_time_in_seconds", also in order
+    /// of addition.
+    using hook = std::function<void(const std::string&, const nlohmann::json&)>;
+
+    namespace
+    {
+        // This is where the function objects are stored, given that they will not be removed or reordered, their index
+        // is what edge_sets' value refers to.
+        std::vector<hook> hooks;
+
+        // Edge set, as in edge-triggered, contains an association between the changed property and an index into hooks.
+        std::unordered_multimap<std::string, size_t> edge_sets;
+    } // namespace
+
+    /// \brief Adds a hook to be run whenever a key on the list of hooks changes.
+    ///        \see hook
+    /// \param [in] hook_triggers An array or container of json paths to match, this
+    ///             container must support begin and end iterators.
+    ///
+    ///             The special value CATCHALL_KEYWORD matches any property that has changed and gets run first
```

Comment:
`CATCHALL_KEYWORD` -> `\p KW_CATCHALL`

The `\p` applies special formatting so that it stands out more. This tells the developer reading the docs that this symbol is special.